### PR TITLE
Add functional test for Cleanup dlq task

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-538

### Change description ###
Duplicate of https://github.com/hmcts/bulk-scan-orchestrator/pull/353
Added functional test to verify if Dlq scheduler task is deleting the messages.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: infrastructure/output.tf [hunk 1]
```diff
@@ -16,6 +16,10 @@ output "ENVELOPES_QUEUE_WRITE_CONN_STRING" {
   value = "${data.terraform_remote_state.shared_infra.queue_primary_send_connection_string}"
 }
 
+output "ENVELOPES_QUEUE_READ_CONN_STRING" {
+  value = "${data.terraform_remote_state.shared_infra.queue_primary_listen_connection_string}"
```
Comment (nitinprabhu) [inline infrastructure/output.tf]: Can we avoid using remote state ? We can use key vault instead

File: Jenkinsfile_CNP [hunk 2]
```diff
@@ -55,6 +55,7 @@ withPipeline(type, product, component) {
 
       // Get envelopes queue connection string
       def sbConnectionStr = kubectl.getSecret(sbNamespaceSecret, namespace, "{.data.connectionString}")
+      env.ENVELOPES_QUEUE_READ_CONN_STRING = "${sbConnectionStr};EntityPath=envelopes"
```
Comment (nitinprabhu) [inline Jenkinsfile_CNP]: same string not sure if we need this ?

Comment (aliveni) [reply Jenkinsfile_CNP]: will remove it

File: src/functionalTest/resources/application.yaml [hunk 3]
```diff
@@ -14,6 +14,7 @@ azure:
   servicebus:
     envelopes:
       max-delivery-count: 10
+      connection-string: ${ENVELOPES_QUEUE_READ_CONN_STRING}
```
Comment (nitinprabhu) [inline src/functionalTest/resources/application.yaml]: This cane be defined after this  
https://github.com/hmcts/bulk-scan-orchestrator/blob/master/src/functionalTest/resources/application.yaml#L8-L10

Comment (aliveni) [reply src/functionalTest/resources/application.yaml]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    @Qualifier("dlqReceiver")
+    private Supplier<IMessageReceiver> dlqReceiverProvider;
+
+    @Autowired
+    private EnvelopeMessager envelopeMessager;
+
+    @Test
+    public void testCleanupDlqTask() throws Exception {
+        // when
+        LOG.info("Send invalid messages to envelopes queue. filename: envelopes/dead-letter-envelope.json");
```
Comment (nitinprabhu) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java]: we should remove this ?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+        // invalid message, which will be sent to dead letter queue
+        for (int i = 0; i < 10; i++) {
```
Comment (nitinprabhu) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java]: can we not send just single message do we need 10 messages ?

File: infrastructure/output.tf [hunk 6]
```diff
@@ -16,6 +16,10 @@ output "ENVELOPES_QUEUE_WRITE_CONN_STRING" {
   value = "${data.azurerm_key_vault_secret.envelopes_queue_send_conn_str.value}"
 }
 
+output "ENVELOPES_QUEUE_READ_CONN_STRING" {
+  value = "${data.azurerm_key_vault_secret.envelopes_queue_listen_conn_str.value}"
```
Comment (nitinprabhu) [inline infrastructure/output.tf]: is the secret config missing here or was already added ?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java (hunk 4 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java]: I added it for debugging. yeah, can be removed

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java (hunk 5 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java]: When we send only one message to the queue, if the dlq task acquires the lock on the message, then the test would pass even though the message still exists in dlq.  
So last week, we decided to send more messages to the queue to make sure the dlq task is running and completing the messages.

File: infrastructure/output.tf (hunk 6 above)
Comment (aliveni) [reply infrastructure/output.tf]: It was already added in the recent pr https://github.com/hmcts/bulk-scan-orchestrator/pull/352

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java (hunk 5 above)
Comment (nitinprabhu) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java]: A comment would be much appreciated :)

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java (hunk 4 above)
Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/CleanupDlqMessagesTest.java]: removed log

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/FunctionalQueueConfig.java [hunk 7]
```diff
@@ -17,19 +22,41 @@
...
+    @Bean(name = "dlqReceiver")
+    public Supplier<IMessageReceiver> dlqReceiverProvider() {
+        return () -> {
+            try {
+                return ClientFactory.createMessageReceiverFromConnectionStringBuilder(
+                    new ConnectionStringBuilder(StringUtils.join(queueReadConnectionString, "/$deadletterqueue")),
+                    ReceiveMode.PEEKLOCK
+                );
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
```
Comment (doncem) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/FunctionalQueueConfig.java]: i know it's just a test config but should the exception be thrown here as well?

Comment (aliveni) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/FunctionalQueueConfig.java]: I'll create a new pr. As discussed, it should be thrown in the main code as well

**Pull request merged**